src/libcrun/net_device.c

```diff
@@ -0,0 +1,451 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * Copyright (C) 2025 Giuseppe Scrivano <giuseppe@scrivano.org>
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#define _GNU_SOURCE
+
+#include <config.h>
+#include "net_device.h"
+#include "utils.h"
+
+#include <sys/socket.h>
+#include <errno.h>
+#include <sched.h>
+#include <sys/wait.h>
+#include <sys/xattr.h>
+#include <linux/netlink.h>
+#include <linux/rtnetlink.h>
+#include <arpa/inet.h>
+#include <net/if.h>
+
+struct ip_addr
+{
+  int len;
+  int ifa_family;
+  int ifa_prefixlen;
+  char data[0];
+};
+
+struct nl_req
+{
+  struct nlmsghdr nlh;
+  union
+  {
+    struct ifaddrmsg ifa;
+    struct ifinfomsg ifi;
+    struct rtgenmsg gen;
+  };
+};
+
+static int
+open_netlink_fd (libcrun_error_t *err)
+{
+  cleanup_close int sock = -1;
+  struct sockaddr_nl local = {
+    .nl_family = AF_NETLINK
+  };
+  int fd;
+
+  sock = socket (AF_NETLINK, SOCK_RAW, NETLINK_ROUTE);
+  if (sock < 0)
+    return crun_make_error (err, errno, "netlink socket");
+
+  if (bind (sock, (struct sockaddr *) &local, sizeof (local)) < 0)
+    return crun_make_error (err, errno, "bind");
+
+  fd = sock;
+  sock = -1;
+  return fd;
+}
+
+/* if_nametoindex with an open netlink socket.  */
+static int
+name_to_index (int sock, const char *ifname, char *buffer, size_t buffer_size, libcrun_error_t *err)
+{
+  struct nl_req req;
+  struct sockaddr_nl sa = { 0 };
+  struct iovec iov = { buffer, buffer_size };
+  struct msghdr msg;
+  int index = 0;
+  int ret;
+
+  memset (&req, 0, sizeof (req));
+
+  req.nlh.nlmsg_len = NLMSG_LENGTH (sizeof (struct rtgenmsg));
+  req.nlh.nlmsg_type = RTM_GETLINK;
+  req.nlh.nlmsg_flags = NLM_F_REQUEST | NLM_F_DUMP;
```

Comment:
actually you don't have to dump all links to get a single one, you can request by name and that is what we are doing in netavark.
https://github.com/containers/netavark/blob/c9649e875a9d5932fad8293095bac2938288a93f/src/network/netlink.rs#L132

This makes things much simpler as you don't have to loop over all links and compare each name yourself.

Reply:
thanks, that was one of the things I wanted to address before moving it out of Draft, plus some other cleanups